src/Container/ApplicationConfigInjectionDelegator.php

```diff
@@ -0,0 +1,373 @@
+            Application::ROUTING_MIDDLEWARE,
+            Application::DISPATCH_MIDDLEWARE,
+        ];
+
+        return function ($item) use ($appMiddleware) {
+            if (in_array($item, $appMiddleware, true)) {
+                return ['middleware' => $item];
+            }
+
+            if (! is_array($item) || ! array_key_exists('middleware', $item)) {
+                throw new InvalidArgumentException(sprintf(
+                    'Invalid pipeline specification received; must be an array'
+                    . ' containing a middleware key; received %s',
+                    is_object($item) ? get_class($item) : gettype($item)
+                ));
+            }
+
+            return $item;
+        };
+    }
+
+    /**
+     * Create reducer function that will reduce an array to a priority queue.
+     *
+     * Creates and returns a function with the signature:
+     *
+     * <code>
+     * function (SplQueue $queue, array $item) : SplQueue
+     * </code>
+     *
+     * The function is useful to reduce an array of pipeline middleware to a
+     * priority queue.
+     *
+     * @return callable
+     */
+    private static function createPriorityQueueReducer()
+    {
+        // $serial is used to ensure that items of the same priority are enqueued
+        // in the order in which they are inserted.
+        $serial = PHP_INT_MAX;
+        return function ($queue, $item) use (&$serial) {
+            $priority = isset($item['priority']) && is_int($item['priority'])
+                ? $item['priority']
+                : 1;
+            $queue->insert($item, [$priority, $serial]);
+            $serial -= 1;
+            return $queue;
+        };
+    }
+
+    /**
+     * Pipe routing middleware into the application.
+     *
+     * Attempts to pull both the router and response interface from the application
+     * container, falling back to reflection of the application instance
+     * when unable to find either one.
+     *
+     * @return void
+     */
+    private static function pipeRoutingMiddleware(Application $application)
+    {
+        $container = $application->getContainer();
+        if ($container->has(RouteMiddleware::class)) {
+            $application->pipe(RouteMiddleware::class);
+            return;
+        }
+
+        $application->pipe(new RouteMiddleware(
+            self::getRouter($application, $container),
+            self::getResponsePrototype($application, $container)
+        ));
+    }
+
+    /**
+     * Pipe dispatch middleware into the application.
+     *
+     * If the DispatchMiddleware is present in the application's container, it
+     * pipes that service. Otherwise, instantiates it directly and pipes it.
+     *
+     * @return void
+     * @throws MissingDependencyException if the RouterInterface service is not
+     *     found when the RouteMiddleware is not present.
+     */
+    private static function pipeDispatchMiddleware(Application $application)
```

Comment:
This method is necessary in v2 to allow `pipeDispatchMiddleware()` and usage of the `DISPATCH_MIDDLEWARE` constant to work; it is not present in v3.